File: Javascript/src/01.AlgorithmicToolbox/Week1ProgrammingChallenges/maxPairwiseProduct.js
/**
 * Maximum Pairwise Product Problem
 *
 * Given a list of integers, return the maximum pairwise product,
 * i.e. product between the two largest distinct (different index) numbers in the list
 * @param {Array.<number>} numbers list of integers
 * @returns maximum pairwise product
 */
export const maxPairwiseProduct = (numbers) => {
  const [largest, secondLargest] = getMaxPair(numbers);
  return largest * secondLargest;
};

export const getMaxPair = (numbers) => {
  return numbers.reduce(
    ([largest, secondLargest], num) => {
      if (num > largest) {
        return [num, largest];
      }
      if (num > secondLargest) {
        return [largest, num];
      }
      return [largest, secondLargest];
    },
    [0, 0]
  );
};

File: Elixir/lib/01.AlgorithmicToolbox/Week1ProgrammingChallenges/max_pairwise_product.ex
defmodule DataStructuresAndAlgorithms.AlgorithmicToolBox.Week1ProgrammingChallenges.MaxPairwiseProduct do
  @moduledoc """
  # Maximum Pairwise Product Problem
  """

  @doc """
  # Maximum Pairwise Product

  Given a `list` of `integers`, return the maximum pairwise product, i.e. product between the two largest distinct (different index) numbers in the list

  ## Examples

  iex> max_pairwise_product([21468, 16859, 82178, 70496, 82939, 44491])
  6_815_761_142

  iex> max_pairwise_product([10, 9, 8, 7, 6])
  90
  """
  @spec max_pairwise_product(list(integer())) :: integer()
  def max_pairwise_product(numbers) do
    {largest, second_largest} =
      Enum.reduce(numbers, {0, 0}, fn number, {largest, second_largest} = max_pair ->
        cond do
          number > largest -> {number, largest}
          number > second_largest -> {largest, number}
          true -> max_pair
        end
      end)

    largest * second_largest
  end
end
